docs/proposals/26-workload-identity.md

```diff
@@ -0,0 +1,517 @@
+`WorkloadIdentity/token` subresource. As the clients will be providing various
+custom information that will be used for the generation of the JWT, yet another
+resource `TokenRequest` in the API group `authentication.gardener.cloud` will be
+used. It is envisioned this resource to contain just metadata for the context
+where the JWT is being used, e.g. shoot or backup entry identifier. Gardener API
+server must verify the provided metadata and it can enhance the JWT with
+additional information derived from the context, for example with information
+for the project and the seed of the shoot cluster. Gardener API can also add
+global information like a garden cluster identity. As `WorkloadIdentity` already
+specifies duration of the token, `TokenRequest` will not feature such field.
+`TokenRequest` resources will never be persisted in the storage layer, the
+generated token will be written in the `.status.token` field and returned to the
+client as response. The expiration timestamp of the token will be also available
+in the status via the `.status.expirationTimestamp` field.
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: TokenRequest
+spec:
+  contextObject: # Optional field, various metadata about context of use of the token
+    apiVersion: core.gardener.cloud/v1beta1
+    kind: Shoot
+    name: foo
+    namespace: garden-local
+    uid: 54d09554-6a68-4f46-a23a-e3592385d820
+status:
+  token: eyJhbGciOiJ....OkBBrVWA # The generated OIDC token
+  expirationTimestamp: 2024-02-09T16:35:02Z
+```
+
+#### Shoot API
+
+Shoot API will not undergo major changes, it will keep referring to
+SecretBindings in the same namespace by their name, however the `SecretBinding`
+API will be extended in the following way to enable usage of `WorkloadIdentity`
+as infrastructure service credentials:
+
+- `SecretBinding.secretRef` field will be made optional and mutable.
+- new optional field `SecretBinding.workloadIdentityRef` will be introduced, it
+  will refer to a `WorkloadIdentity` resource by its name and namespace. The
+  value of the field will not be immutable and will allow existing shoots to
+  change the `WorkloadIdentity` they are using to allow rotation by the Gardener
+  users.
+- validation will ensure that either `secretRef` or `workloadIdentityRef` is
+  set, but not both.
+- on update of the `workloadIdentityRef`, a validation will ensure that both the
+  old and the new `WorkloadIdentity` are for the same cloud provider account.
+
+From user experience point of view, now `SecretBinding` might not be the best
+name for this resource, as it is no longer limited to referring only secrets, as
+its name implies. In a future version of Gardener APIs, it could and would be
+nice to be renamed to `CredentialsConfig`, `CredentialsBinding`,
+`CredentialsSource` or something similar. This way it also leave open space for
+other credentials implementations in the future.
```

Comment:
It would have been easier if the API was still in alpha version, however now it will require long deprecation periods, intensive communication with users, changes in a lot of tools integrated with Gardener. It will unnecessary and artificially slow down development and rollout of workload identity, therefore it is better to be done separately when the API is redesigned, eventually in the future.

Reply:
Why not add a separate option in the shoot API next to `.spec.secretBindingName`, e.g. `.spec.credentialBindingName`? It is fairly easy/straight-forward to validate that only one of the two is set. The new API could be introduced directly and the old API could still stay in place for a longer period of time. Eventually, we could decide to either remove `.spec.secretBindingName` or have essentially two different options how credentials could be provided.
WDYT?

Reply:
I was also thinking in this direction, but at end it looks like a bit much work and out of the scope of this GEP. And mostly the coordination with the different clients dealing with SecretBindings will imply unwanted artificial delays.

Reply:
Maybe I did not understand you fully, but we need to anyway perform some changes as the credential handling is slightly different. If we go forward with the approach described in the GEP and store the credentials again in the `cloudprovider` secret in the control plane there are no changes on the credential consumer side required. However, this is regardless whether we "extend" `SecretBinding` or use `CredentialBinding` instead. Therefore, I do not get where artificial delays would come in.

For sure, the eventual deprecation/removal of the old API would take time, but this is true regardless if the change is done now or later.

I would not want the `SecretBinding` API to be extended in a hard to understand way if we just want to take shortcuts. Clean API design should be part of a GEP even if it means that the full implementation of the GEP will take more time.

Reply:
Done with 08fec71e60, please take a look.